readme.md
# M3wP Yahtzee!

## Introduction
Hello and welcome to the **M3wP Yahtzee!** read me!

I have developed a multi-platform client/server version of the Hasbro game Yahtzee!  It is currently in development and quite crude but should be playable on PC, Android, iOS and Mac.  The client will eventually even be available for the humble Commodore 64 and compatibles (because I love that platform).

I now have a Linux CLI server!   There is a CLI server for Win64, as well.

At this point in time, I am only going to be providing binaries for Linux, Windows and Android because I don't have the required development tools for iOS and Mac.  The C64 client is now available too but is still unfinished.

## Download and Installation
The Windows x86 (32 bit) binaries are available in the _bin/Win32_ folder as compressed files.  You will need to decompress them before you can run them.  Likewise, the Windows x64 binaries are in the _bin/Win64_ folder.

The Android client binary (I have not compiled or tested the server on this platform) is available in the _bin/Android_ folder.  It is currently compiled only for debugging and you will need to manually install it using adb.

You'll find the Linux x64 CLI server in _bin/Linux64_.

The Commodore64 binaries are available in the _bin/C64_ folder.

## Compiling
The Windows, Android, iOS and Mac versions require Delphi FMX.  I am using Delphi XE8 at the present time.  You should find all you need in the _src/FMX_ folder.

You could compile the FMX client and server to Linux if you have the appropriate version of Delphi.  If you can do this, please let me know and forward me the binaries.  I'd love to hear from you!

The CLI server is a bit tricky.  You will need the Areat Synapse library in your _src_ tree.  Download it and decompress it to _src/synalist-code_.

The C64 version is a bit more complicated.  You need the cc65 toolchain, the ip65 library and some makefiles.  I will document this further and provide more detail when the client matures a little more.

## Playing the Game
You will need to have a server running somewhere.  Then, get your friends together and each of you fire up an instance of the client.

You will need to supply the Host address or name and a unique User Name (which should be no more than 8 characters long, with no spaces and only ASCII characters).  Clicking "Connect" will connect your client to the server.  If you have connection difficulties, it is most likely that your User Name is incorrect or already in use.  Try changing it and clicking the "Update" button.

Once connected, you will be in the Lobby.  You can create Chat "Rooms" from where you can talk to the other connected users from the Chat tab.  Try using the "List" button to show the currently available Rooms.  Type in the name of a Room to join and click "Join" to enter that room.  Use the navigation buttons to change between the Room conversation and control tabs.  Click the "Members" button to show the users presently in the same Room.  Clicking the "Part" button makes you leave the Room.

When you are ready, go to the Play tab where you can create and list Games much as you do with Rooms.  Join a Game like you do with a Room and navigate to the Overview tab with the navigation buttons.

You will need to inform the rest of the Players that you are Ready to play by clicking the Ready button.  If you are waiting for other Players, don't click it yet.  When everyone is Ready, you must Roll For First by clicking the button (which has now changed).  New Players cannot join the game at this point.  The Player with the highest roll will go first and the icons will show that they have the dice and that the other Players are now waiting.

Clicking on a Player will show that Player's Score Card.  If that Player is the one with the dice, you can click the "Follow Active" check box to always show the player with the dice.  This is recommended if you want to follow the game play.  Alternatively, use the navigation buttons to show the desired information.

If it is your turn, you must now click on the "Roll" button to roll the dice.  Click on each Die to mark it as a "Keeper" or not.  Keepers are shown below the ones that will be re-rolled if the Roll button is clicked again.  You can Roll Dice up to 3 times.

Once you are finished you must place a score on your Score Card.  Clicking a Score Cell on the Score Card will show you what score you will get for that position.  Chose wisely!  When you are satisfied with your selection, click the "Score" button and the Dice will be passed to the next Player.

Play continues in this fashion until each player has filled their Score Card for a total of 13 rounds.  The Winner is the Player with the highest Score.

When you are finished playing, click the "Part" button to leave the game.  You are ready to start again!

**Note:**  The C64 client usage will be a little different to this but you should get the idea.

## Future Development
There are many planned features for the game and the apps (both client and server) will be enhanced to be prettier and show more information.  The project is currently in an Alpha state but should be playable for the most part.

I am in the process of rebuilding the server and client using Areat Synapse and LCL instead of Indy10 and FMX.  I have seriously had enough of the problems involved.  The FMX client will be maintained essentially for Android and iOS only.

## Contact
Please feel free to contact me for further information.  You can reach me at:
	
	mewpokemon {at} hotmail {dot} com

Please include the string "M3wPYahtzee!" in the subject line or your mail might get lost.

changes.md
# M3wP Yahtzee! Change Log

## CLI Server

* 06NOV2019
	* Fix issue releasing a Connection via DisconnectByIdent

* 03NOV2019
	* Plug potential holes in logic updating a player user name, creating a new room and creating a new game

* 01NOV2019
	* Reinstate software challenges to client to keep alive
	* Revert changes for checking connection and keep-alives - wasn't working according to documents I had read

* 30OCT2019
	* Implement hash bucket workers instead of creating a thread for each connection
	* Check for disconnects according to standards (will also not accept OOB data!)
	* Implement connection "keep-alives" according to standards (Windows, need Linux)
	* Store the connection "ticket" on the player, too
	* Use connection "ticket" for output instead of ident (easier to read)
	* Generate a "ticket" for each connection, based on its "ident" (prepare for client "re-auths")

* 29OCT2019
	* Yahtzee Bonuses only apply when you haven't got a Yahtzee score of 0 and you can only have three of them!

* 28OCT2019
	* Incorporate new TCPServer into code
	* Implement my own TCPServer to handle multiple connections based on Ararat Synapse
	* REMOVE INDY10!  I'm over it.  Too many problems.

* 25OCT2019
	* Work around Indy sometimes not firing OnDisconnect when a connection is "abandonded"
	* Implement ExiprePlayers functionality to ensure release of memory despite behaviour of Connection disconnect - affects all servers
	* Fix bug in PlayersKeepAliveExpire to call correct Remove (of Self) - affects all servers

* 23OCT2019
	* Make CPU usage nicer?

* 22OCT2019
	* Also use the server challenges

* 21OCT2019
	* Fix a number of memory leaks (affects the FMX server, too)


## LCL Client

### Version 0.00.80A

* 01NOV2019
	* Implement LCL client for native Linux solution and testing comparisson


## FMX Server

### Version 0.00.78A

* 29OCT2019
	* Yahtzee Bonuses only apply when you haven't got a Yahtzee score of 0 and you can only have three of them!
	
* 25OCT2019
	* Work around Indy sometimes not firing OnDisconnect when a connection is "abandonded"

* 22OCT2019
	* Fix some FIXMEs (affects all servers)
	* Implement server challenge message when client idle

* 17OCT2019
	* Handle when the currently playing player leaves the game

* 16OCT2019
	* Change Play|Join message to be all text, not mixed text and binary
	* Fix exception when closing if a player is in a game/room
	* Change list message response to Play|List to be all text, not mixed text and binary


## FMX Client

### Version 0.00.78A

* 22OCT2019
	* Implement client keep-alive message in response to server challenge message

* 16OCT2019
	* Change Play|Join message to be all text, not mixed text and binary
	* Change list message handling of Play|List response to expect all text, not mixed text and binary


## C64 Client

### Version 0.00.35A

* 02NOV2019
	* Allow the user to update their user name (still not updating back from server)
	* Don't change the active control when parting a room if in less mode
	* Don't part the room unless we were the one parting
	* Change "their" to "this" score on Detail Page.
	* Clear "their" score when showing your details.
	* Make sure the selection can be properly cleared in the Score Sheet Panel when selecting aces
	* Change ip65 such that it no longer spin waits for no reason on tcp_close
	* Should now be able to try to connect when a connection attempt fails?
	* Flash notify and (when appropriate) disconnect if fail to get new internet send message
	* Provide failure feedback when attempting to get a new internet send message
	* Implement flash notification functionality
	* Use send data metrics to produce a connection health meter
	* Change ip65 to export send data metrics

* 30OCT2019
	* Fixes to ip65 for reading while sending causing disconnects (require modified ip65)
	* Fixes to ip65 for timeouts when sending (require modified ip65)
	* Update to latest ip65 version
	* Work-around inet message send queue overflow

* 26OCT2019
	* Default host name to "PLAY-CLASSICS.NET"!
	* Implement Lobby|Peer message receive
	* Panels must be enabled and visible when finding an element key accelerator
	* Implement Lobby|Join and Lobby|Part functionality (UI and messages)
	* Implement y offset in log panel control
	* (Hopefully) fix default active control assignment in prepare procedure
	* Always prepare all controls on a page (regardless of panel visibility)

* 25OCT2019
	* Actually allocate a byte for the score sheet's "hveprvw" field
	* Try to update the score sheet's selection indicator more accurately
	* Don't update the dice when Play|KeeperPeer message unless viewing the slot's details
	* Refactor clientProcPlayStatGameMsg and clientUpdateSlotState
	* Update game round on Play|Overview page as game plays
	* Update "our" and "their" scores on Play|Detail page as game plays
	* Handle Play|ScoreQuery message send and receive
	* Implement score query display (still need yahtzee/lower bonus display)

### Version 0.00.33A

* 25OCT2019
	* Now show roll number on Play|Detail page roll button
	* Activate Play|Detail roll button when current player gets play status
	* Update Play|Detail player label correctly
	* Output error string for Play|Error messages
	* Play|StatusPeer message handling should be more "reliable" now

* 24OCT2019
	* Fix label accelerators not activate/down "attached" controls

* 20OCT2019
	* Bug fixes for displaying scores on the Play|Detail page
	* Handle send Play|ScorePeer message
	* Handle receive Play|ScorePeer message
	* Display scores from selected detail slot on score sheet control (need yahtzee bonus handling)
	* Further work on Play|Detail page updating
	* Implement Play|KeepersPeer message send and receive
	* Implement Play|Roll message send and receive
	* Update the Play|Detail page when switched to and when receive the Play|StatPeer messages

### Version 0.00.23A

* 20OCT2019
	* Implement OPT_AUTOCHECK functionality for checkbox controls

src/YahtzeeClasses.pas
unit YahtzeeClasses;

{$IFDEF FPC}
	{$MODE DELPHI}
{$ENDIF}
{$H+}

interface

uses
	Generics.Collections, Classes;

type
 	TMsgData = array of Byte;

     { TBaseIdentMessage }

    TBaseIdentMessage = class
		Ident: TGUID;
        Data: TMsgData;

        constructor Create; virtual;

		function  Encode: TMsgData; virtual;
		procedure Decode(const AData: TMsgData); virtual;
	end;

    TIdentMessages = TThreadList<TBaseIdentMessage>;

    TMsgCategory = (mcSystem, mcText, mcLobby, mcConnect, mcClient, mcServer, mcPlay);

	TBaseMessage = class(TBaseIdentMessage)
	public
		Category: TMsgCategory;
		Method: Byte;
		Params: TList<AnsiString>;

		constructor Create; override;
		destructor  Destroy; override;

		procedure ExtractParams; virtual;
		procedure DataFromParams;

		function  DataToString: AnsiString;

		function  Encode: TMsgData; override;
		procedure Decode(const AData: TMsgData); override;

		procedure Assign(AMessage: TBaseMessage);
	end;

//	TMessages = TThreadList<TMessage>;

    TLogKind = (slkError, slkWarning, slkInfo, slkDebug);

    TLogMessage = class
        Kind: TLogKind;
        Message: string;
    end;

	TLogMessages = TThreadList<TLogMessage>;

	TNamedHost = class(TObject)
	public
		Name: AnsiString;
		Host: AnsiString;
		Version: AnsiString;
	end;

	TGameState = (gsWaiting, gsPreparing, gsPlaying, gsPaused, gsFinished);
	TPlayerState = (psNone, psIdle, psReady, psPreparing, psWaiting, psPlaying,
			psFinished, psWinner);

	TDie = 0..6;
	TDice = array[0..4] of TDie;
	TDieSet = set of 1..5;

	TScoreLocation = (slAces, slTwos, slThrees, slFours, slFives, slSixes,
			slUpperBonus, slThreeKind, slFourKind, slFullHouse, slSmlStraight,
			slLrgStraight, slYahtzee, slChance, slYahtzeeBonus1, slYahtzeeBonus2,
			slYahtzeeBonus3);
	TScoreLocations = set of TScoreLocation;

	TScoreSheet = array[TScoreLocation] of Word;

	function  DieSetToByte(ADieSet: TDieSet): Byte;
	function  ByteToDieSet(AByte: Byte): TDieSet;
	procedure RollDice(ASet: TDieSet; var ADice: TDice);

	function  MakeScoreForLocation(ALocation: TScoreLocation; ADice: TDice;
			var AUsed: TDieSet): Word;

	function  IsYahtzee(ADice: TDice): Boolean;
	function  IsYahtzeeBonus(ASheet: TScoreSheet; var ALocation: TScoreLocation): Boolean;
	function  YahtzeeBonusStealLocs(ASheet: TScoreSheet; ADice: TDice): TScoreLocations;
	function  YahtzeeBonusStealScore(ALocation: TScoreLocation; ADice: TDice): Word;

const
	VAL_KND_SCOREINVALID = High(Word);

	ARR_LIT_NAM_CATEGORY: array[TMsgCategory] of string = (
			'system', 'text', 'lobby', 'connect', 'client', 'server', 'play');

	VAL_SET_DICEALL: TDieSet = [1..5];

	VAL_SET_SCOREUPPER: TScoreLocations = [slAces..slSixes];
	VAL_SET_SCORELOWER: TScoreLocations = [slThreeKind..High(TScoreLocation)];

	ARR_LIT_NAME_SCORELOC: array[TScoreLocation] of string = (
			'Aces (1''s)', 'Twos (2''s)', 'Threes (3''s)', 'Fours (4''s)',
			'Fives (5''s)', 'Sixes (6''s)', 'Upper Bonus', '3 of a Kind',
			'4 of a Kind', 'Full House', 'SM Straight', 'LG Straight', 'YAHTZEE',
			'Chance', 'YAHTZEE Bonus', 'YAHTZEE Bonus', 'YAHTZEE Bonus');


//  I think that 2 should be 5 but this is using RFC messages as a template.

//		0	-	System
//		00	- 	Hang up
//		0E	-	Invalid category
//		0F	-	Invalid empty
//
//		1	-	Text
//		10	-	Information
//		11	-	Begin
//		12	-	More
//		13	-	Data
//		14	-	Peer
//
//		2	-	Lobby
//		20 	- 	Error
//		21	-	Join
//		22	-	Part
//		23	-	List
//		24	-	Peer
//
//		3	-	Connection
//		30	-	Error
//		31	-	Identify
//
//		4	-	Client
//		40	-	Error
//		41	-	Identify
//		52	-	KeepAlive
//
//		5	-	Server
//		50	-	Error
//		51	-	Identify
//		52	-	Challenge
//
//		6	-	Play
//		60 	- 	Error
//		61	-	Join
//		62	-	Part
//		63	-	List
//		64	-	TextPeer
//		65	-	KickPeer
//		66	-	StatusGame
//		67	-	StatusPeer
//		68	-	RollPeer
//		69	-	KeepersPeer
//		6A	-	ScoreQuery
//		6B	-	ScorePeer

procedure AddLogMessage(const AKind: TLogKind; const AMessage: string);

var
	LogMessages: TLogMessages;


implementation

uses
    SysUtils;


procedure RollDice(ASet: TDieSet; var ADice: TDice);
	var
	i: Integer;

	begin
	for i:= 1 to 5 do
		if  i in ASet then
			ADice[i - 1]:= Random(6) + 1;
	end;


procedure AddLogMessage(const AKind: TLogKind; const AMessage: string);
    var
    lm: TLogMessage;

    begin
    lm:= TLogMessage.Create;
    lm.Kind:= AKind;
    lm.Message:= FormatDateTime('hh:nn:ss.zzz ', Now) + AMessage;
    UniqueString(lm.Message);

    LogMessages.Add(lm);
    end;

function  DieSetToByte(ADieSet: TDieSet): Byte;
	var
	i: TDie;
	b: Byte;

	begin
	Result:= 0;
	b:= $01;

	for i:= 1 to 5 do
		begin
		if  i in ADieSet then
			Result:= Result or b;
		b:= b shl 1;
		end;
	end;

function  ByteToDieSet(AByte: Byte): TDieSet;
	var
	i: TDie;
	b: Byte;

	begin
	Result:= [];
	b:= $01;

	for i:= 1 to 5 do
		begin
		if  (b and AByte) <> 0 then
			Include(Result, i);
		b:= b shl 1;
		end;

	end;


function TallyDieScoreFor(ADie: Integer; ADice: TDice; var AUsed: TDieSet): Word;
	var
	i: Integer;

	begin
	Result:= 0;

	for i:= 0 to 4 do
		if  ADice[i] = ADie then
			begin
			Include(AUsed, i + 1);
			Inc(Result, ADie);
			end;
	end;

function TallyAllDieScore(ADice: TDice): Word;
	var
	i: Integer;

	begin
	Result:= 0;

	for i:= 0 to 4 do
		Inc(Result, ADice[i]);
	end;

function  MakeScoreForLocation(ALocation: TScoreLocation; ADice: TDice;
		var AUsed: TDieSet): Word;

	function HaveCountDie(ACount: Integer; ADice: TDice; var AUsed: TDieSet): Boolean;
		var
		i,
		j: Integer;
		d,
		c: Integer;
		u: TDieSet;

		begin
		Result:= False;

		for i:= 0 to 4 do
			if  not ((i + 1) in AUsed) then
				begin
				d:= ADice[i];
				u:= AUsed;
				c:= 0;

				for j:= 0 to 4 do
					if  not ((j + 1) in u) then
						if d = ADice[j] then
							begin
							Inc(c);
							Include(u, j + 1);
							if  c = ACount then
								Break;
							end;

				if  c = ACount then
					begin
					Result:= True;
					AUsed:= u;
					Break;
					end;
				end;
		end;

	function TallyDieScoreKind(ACount: Integer; ADice: TDice; var AUsed: TDieSet): Word;
		var
		i: Integer;
		f: Boolean;

		begin
		Result:= 0;

		if  ACount > 0 then
			begin
			f:= False;
			for i:= 0 to 4 do
				if  HaveCountDie(ACount, ADice, AUsed) then
					begin
					f:= True;
					Break;
					end;
			end
		else
			f:= True;

		if  f then
			Result:= TallyAllDieScore(ADice);
		end;

	function FindFullHouse(ADice: TDice; var AUsed: TDieSet): Word;
		begin
		Result:= 0;

		if  HaveCountDie(3, ADice, AUsed)
		and HaveCountDie(2, ADice, AUsed) then
			Result:= 25;
		end;

	function HaveSequence(AStart, ASize: Integer; ADice: TDice; var AUsed: TDieSet): Boolean;
		var
		u: TDieSet;
		c: Integer;
		i,
		j: Integer;
		d: Integer;
		f: Boolean;

		begin
		Result:= False;

		for i:= 0 to 4 do
			begin
			d:= AStart;
			u:= AUsed;
			c:= 0;

			if  not ((i + 1) in AUsed) then
				if  ADice[i] = d then
					begin
					f:= True;
					Inc(c);
					Include(u, i + 1);
					Dec(d);

					while f do
						begin
						f:= False;

						for j:= 0 to 4 do
							if  (not ((j + 1) in AUsed))
							and (ADice[j] = d) then
								begin
								Inc(c);
								f:= c < ASize;
								Dec(d);
								Include(u, j + 1);
								Break;
								end;

						end;

					if  c = ASize then
						begin
						Result:= True;
						AUsed:= u;
						Break;
						end;
					end;
			end;
		end;

	function FindStraight(ASize: Integer; ADice: TDice; var AUsed: TDieSet): Word;
		var
		i: Integer;
		n: Integer;

		begin
		Result:= 0;

		n:= 6 - (6 - ASize);

		for i:= 6 downto n do
			if  HaveSequence(i, ASize, ADice, AUsed) then
				begin
				if  ASize = 5 then
					Result:= 40
				else
					Result:= 30;
				Break;
				end;
		end;

	begin
//	Result:= 0;
	AUsed:= [];

	if  ALocation in VAL_SET_SCOREUPPER then
		Result:= TallyDieScoreFor(Ord(ALocation) + 1, ADice, AUsed)
	else
		case ALocation of
			slThreeKind:
				Result:= TallyDieScoreKind(3, ADice, AUsed);
			slFourKind:
				Result:= TallyDieScoreKind(4, ADice, AUsed);
			slFullHouse:
				Result:= FindFullHouse(ADice, AUsed);
			slSmlStraight:
				Result:= FindStraight(4, ADice, AUsed);
			slLrgStraight:
				Result:= FindStraight(5, ADice, AUsed);
			slYahtzee:
				if  HaveCountDie(5, ADice, AUsed) then
					Result:= 50
				else
					Result:= 0;
			slChance:
				begin
				Result:= TallyAllDieScore(ADice);
				AUsed:= VAL_SET_DICEALL;
				end;
			slYahtzeeBonus1..slYahtzeeBonus3:
				if  HaveCountDie(5, ADice, AUsed) then
					Result:= 100
				else
					Result:= 0;
			else
				Result:= VAL_KND_SCOREINVALID;
			end;
	end;

function  IsYahtzee(ADice: TDice): Boolean;
	var
	i: Integer;
	d: Integer;

	begin
	Result:= True;

	d:= ADice[0];
	for i:= 1 to 4 do
		if  ADice[i] <> d then
			begin
			Result:= False;
			Break;
			end;
	end;

function  IsYahtzeeBonus(ASheet: TScoreSheet; var ALocation: TScoreLocation): Boolean;
	var
	i,
    j: TScoreLocation;


	begin
	Result:= False;

	if  (ASheet[slYahtzee] <> VAL_KND_SCOREINVALID)
    and (ASheet[slYahtzee] <> 0) then
		begin
//		Result:= True;

        j:= slAces;

		for i:= slYahtzeeBonus1 to slYahtzeeBonus3 do
			if  ASheet[i] = VAL_KND_SCOREINVALID then
				begin
				j:= i;
				Break;
				end;

        Result:= j > slAces;

        if  Result then
            ALocation:= j;
		end;
	end;

function  YahtzeeBonusStealLocs(ASheet: TScoreSheet; ADice: TDice): TScoreLocations;
	var
	i: TScoreLocation;

	begin
	Result:= [];

	if  ASheet[TScoreLocation(ADice[0] - 1)] = VAL_KND_SCOREINVALID then
		Result:= [TScoreLocation(ADice[0] - 1)]
	else
		begin
		if  ASheet[slThreeKind] = VAL_KND_SCOREINVALID then
			Include(Result, slThreeKind);

		if  ASheet[slFourKind] = VAL_KND_SCOREINVALID then
			Include(Result, slFourKind);

		if  Result = [] then
			begin
			for i:= slFullHouse to slLrgStraight do
				if  ASheet[i] = VAL_KND_SCOREINVALID then
					Include(Result, i);

			if  ASheet[slChance] = VAL_KND_SCOREINVALID then
				Include(Result, slChance);
			end;
		end;

	if  Result = [] then
		for i:= slAces to slSixes do
			if  ASheet[i] = VAL_KND_SCOREINVALID then
				Include(Result, i);
	end;

function  YahtzeeBonusStealScore(ALocation: TScoreLocation; ADice: TDice): Word;
	var
	u: TDieSet;

	begin
	u:= [];

	if  ALocation in [slAces..slSixes] then
		Result:= TallyDieScoreFor(Ord(ALocation) + 1, ADice, u)
	else if ALocation in [slThreeKind..slFourKind, slChance] then
		Result:= TallyAllDieScore(ADice)
	else if ALocation = slFullHouse then
		Result:= 25
	else if ALocation = slSmlStraight then
		Result:= 30
	else if ALocation = slLrgStraight then
		Result:= 40
	else
		Result:= 0;
	end;

{ TBaseIdentMessage }

constructor TBaseIdentMessage.Create;
    begin
    inherited;

    end;

function TBaseIdentMessage.Encode: TMsgData;
    begin
    Result:= Data;
    end;

procedure TBaseIdentMessage.Decode(const AData: TMsgData);
    begin
    Data:= AData;
    end;


{ TMessage }

procedure TBaseMessage.Assign(AMessage: TBaseMessage);
	var
	i: Integer;

	begin
	Category:= AMessage.Category;
	Method:= AMessage.Method;

	SetLength(Data, Length(AMessage.Data));
	Move(AMessage.Data[0], Data[0], Length(AMessage.Data));

	Params.Clear;
	for i:= 0 to AMessage.Params.Count - 1 do
		Params.Add(AMessage.Params[i]);
	end;

constructor TBaseMessage.Create;
	begin
	inherited Create;

	Params:= TList<AnsiString>.Create;
	end;

procedure TBaseMessage.DataFromParams;
	var
	s: AnsiString;
	i: Integer;

	begin
	s:= AnsiString('');
	for i:= 0 to Params.Count - 1 do
		begin
		s:= s + Params[i];
		if  i < (Params.Count - 1) then
			s:= s + AnsiString(' ');
		end;

	SetLength(Data, Length(s));
	for i:= Low(s) to High(s) do
		Data[i - Low(s)]:= Ord(s[i]);
	end;

function TBaseMessage.DataToString: AnsiString;
	var
	i: Integer;

	begin
	Result:= AnsiString('');
	for i:= 0 to High(Data) do
		Result:= Result + AnsiChar(Data[i]);
	end;

procedure TBaseMessage.Decode(const AData: TMsgData);
	var
	i: Integer;
	c: Byte;

	begin
	if  (Length(AData) > 0)
	and (Length(AData) = AData[0] + 1) then
		begin
		SetLength(Data, Length(AData) - 2);

		c:= AData[1] shr 4;
		if  c in [Ord(Low(TMsgCategory))..Ord(High(TMsgCategory))] then
			begin
			Category:= TMsgCategory(AData[1] shr 4);
			Method:= AData[1] and $0F;
			end
		else
			begin
			Category:= mcSystem;
			Method:= $0E;
			end;

		for i:= 2 to High(AData) do
			Data[i - 2]:= AData[i]
		end
	else
		begin
		SetLength(Data, 0);
		Category:= mcSystem;
		Method:= $0F;
		end;
	end;

destructor TBaseMessage.Destroy;
    var
    s: AnsiString;

    begin
    with Params do
        while Count > 0 do
            begin
            s:= Items[0];
            Delete(0);
            end;

	Params.Free;

	inherited;
	end;

function TBaseMessage.Encode: TMsgData;
	var
	i: Integer;
	c: Byte;

	begin
	SetLength(Result, 2 + Length(Data));

	Result[0]:= Length(Data) + 1;

	c:= (Ord(Category) shl 4) or (Method and $0F);
	Result[1]:= c;

	for i:= 0 to High(Data) do
		Result[2 + i]:= Data[i];
	end;

procedure TBaseMessage.ExtractParams;
	var
	i: Integer;
	s: AnsiString;

	begin
    with Params do
        while Count > 0 do
            begin
            s:= Items[0];
            Delete(0);
            s:= AnsiString('');
            end;

//	Params.Clear;

    s:= AnsiString('');
	for i:= 0 to High(Data) do
		if  Data[i] = $20 then
			begin
			Params.Add(s);
			s:= AnsiString('');
			end
		else
			s:= s + AnsiChar(Data[i]);

	if  Length(s) > 0 then
		Params.Add(s);
	end;


initialization
	LogMessages:= TLogMessages.Create;


finalization
    with LogMessages.LockList do
        while Count > 0 do
            begin
            Items[Count - 1].Free;
            Delete(Count - 1);
            end;

	LogMessages.Free;

end.
